### PR TITLE
v0.21: Facets and filters

### PR DESCRIPTION
**Pull request by guimachiavelli**
This PR:

- [x] creates a new page under `reference/features`: `Filtering and faceted search`;
- [x] removes `reference/features/filtering.md`: content has been moved to new `Filtering and faceted search` page;
- [x] removes `reference/features/faceted_search.md`: content has been moved to new `Filtering and faceted search` page;
- [x] updates `reference/api/search_parameters.md`: remove `filters` and `facetFilters`, add `filter`, update `facetsDistribution`;
- [x] updates `reference/api/settings.md`: rename `attributesForFaceting` => `filterableAttributes`;
- [x] rename `reference/api/attributes_for_faceting.md`: `attributesForFaceting` => `filterableAttributes`;
- [x] fixes broken links caused by setting name change and page removal.

Once this PR is approved:
- [ ] list unused faceted search examples so we can remove them from the code samples.

### CODE REVIEW

File: reference/api/filterable_attributes.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+```json
+["genres", "director"]
+```
+
+## Update filterable attributes
+
+<RouteHighlighter method="POST" route="/indexes/:index_uid/settings/filterable-attributes" />
+
+Update an index's [filterable attributes list](/reference/features/filtering_and_faceted_search.md). This is a slow process that will take as much time as indexing the dataset.
```
Comment (gmourier) [inline reference/api/filterable_attributes.md:37]: Thanks @guimachiavelli, maybe we should also inform the user that he can specify them before indexing the documents if he already knows the fields on which he will want to filter or facet. This way, it avoids having to wait for this second automatic indexing phase.

File: reference/features/filtering_and_faceted_search.md [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    …
+]
+```
+
+If you want to enable filtering using `director`, `release_date`, `genres`, and `rating`, you must add these attributes to the [`filterableAttributes` index setting](/reference/api/filterable_attributes.md).
+
+You can then restrict a search so it only returns movies released after 18 March 1995 with the following filter containing a single condition:
+
+```SQL
+release_data > 795484800
```
Comment (gmourier) [inline reference/features/filtering_and_faceted_search.md:189]:
```suggestion
release_date > 795484800
```

File: reference/features/search_parameters.md [hunk 3]
```diff
@@ -120,107 +119,69 @@ If you want your query to return only **two** documents, set `limit` to `2`:
...
+:::
 
 ### Example
 
-Suppose you are searching for movies, but only want results in `Horror` or `Mystery` `genres` and whose `director` is `Jordan Peele`:
+You can write a filter expression in string syntax using logical connectives:
 
 ```SQL
-("genres:Horror" OR "genres:Mystery") AND "director:Jordan Peele"
+("genres = horror OR genres = mystery) AND director = 'Jordan Peele']"
```
Comment (gmourier) [inline reference/features/search_parameters.md:141]:
```suggestion
"(genres = horror OR genres = mystery) AND director = 'Jordan Peele'"
```

**Review by gmourier — CHANGES_REQUESTED** [review]

File: learn/what_is_meilisearch/features.md [hunk 4]
```diff
@@ -32,15 +32,15 @@ Read more about synonyms in [this dedicated guide](/reference/features/synonyms.
...
+[Faceted search](/reference/features/filtering_and_faceted_search.md) allows you to classify search results into categories and to build intuitive navigation interfaces.
 
 ## Placeholder search
 
-If you make a search without inputting any query words, MeiliSearch will return all the documents in that index sorted by its [custom ranking rules](/reference/features/settings.md#custom-ranking-rule). This feature is called **placeholder search**. It is particularly effective when used with other features such as [faceting or filtering](/reference/features/faceted_search.md#filters-or-facets), which allow users to narrow their searches and browse by category.
+If you make a search without inputting any query words, MeiliSearch will return all the documents in that index sorted by its [custom ranking rules](/reference/features/settings.md#custom-ranking-rule). This feature is called **placeholder search**. It is particularly effective when used with other features such as [faceting or filtering](/reference/features/filtering_and_faceted_search.md#filters-or-facets), which allow users to narrow their searches and browse by category.
```
Comment (bidoubiwa) [inline learn/what_is_meilisearch/features.md:43]:
```suggestion
If you make a search without inputting any query words, MeiliSearch will return all the documents in that index sorted by its [custom ranking rules](/reference/features/settings.md#custom-ranking-rule). 

This feature is called **placeholder search**. It is particularly effective when used with other features such as [faceting or filtering](/reference/features/filtering_and_faceted_search.md#filters-or-facets), which allow users to narrow their searches and browse by category.
```

Should this be a note? Seems like it is important knowledge to understand placeholder search when it is in the same paragraph but it is more a suggestion.

File: reference/api/filterable_attributes.md (hunk 1 above)
Comment (bidoubiwa) [reply reference/api/filterable_attributes.md:37]: Unfortunately this is not explained anywhere in the doc (unless I m mistaken). Maybe we can rephrase it 

```suggestion
Update an index's [filterable attributes list](/reference/features/filtering_and_faceted_search.md) re-indexes all documents in the index.  We suggest adding the settings after index creation, thus before adding your first documents.
```

File: reference/api/search.md [hunk 5]
```diff
@@ -28,15 +28,14 @@ This is the preferred route to perform search when an API key is required, as it
...
 | **limit**                 | `Integer`              | Maximum number of documents returned                                                            | `20`                                                                                            |
-| **filters**               | `String`               | Filter queries by an attribute value                                                            | `null`                                                                                          |
-| **facetFilters**          | `[Strings OR [Strings]]` | Facet names and values to filter on                                                            | `null`                                                                                          |
+| **filter**               | `String`               | Filter queries by an attribute value                                                            | `null`                                                                                          |
```
Comment (bidoubiwa) [inline reference/api/search.md:31]:
```suggestion
| **filter**               | `String` OR [Strings OR [Strings]] | Filter queries by an attribute value                                                            | `null`                                                                                          |
```

`filter` can also be the type of `facetFilters` (a bit chaotic but accurate)

File: reference/api/search.md [hunk 6]
```diff
@@ -28,15 +28,14 @@ This is the preferred route to perform search when an API key is required, as it
...
 | **cropLength**            | `Integer`              | Length used to crop field values                                                                | `200`                                                                                           |
 | **attributesToHighlight** | `[Strings]`            | Attributes whose values will contain highlighted matching terms                                 | `null`                                                                                          |
 | **matches**               | `Boolean`              | Defines whether an object that contains information about the matches should be returned or not | `false`                                                                                         |
-> `filters` accept a query string. You can find more about the filter syntax on [our dedicated page](/reference/features/filtering.md).
+> `filter` accepts a query string. You can find more about the filter syntax on [our dedicated page](/reference/features/filtering_and_faceted_search.md).
```
Comment (bidoubiwa) [inline reference/api/search.md:38]: Not sure I understand this sentence

File: reference/api/search.md [hunk 7]
```diff
@@ -123,16 +122,15 @@ This route should only be used when no API key is required. If an API key is req
...
 | **[attributesToCrop](/reference/features/search_parameters.md#attributes-to-crop)**           | Attributes whose values have to be cropped                                                      |    `null`     |
 | **[cropLength](/reference/features/search_parameters.md#crop-length)**                        | Length used to crop field values                                                                |    `200`      |
 | **[attributesToHighlight](/reference/features/search_parameters.md#attributes-to-highlight)** | Attributes whose values will contain highlighted matching terms                                 |    `null`     |
 | **[matches](/reference/features/search_parameters.md#matches)**                               | Defines whether an object that contains information about the matches should be returned or not |   `false`     |
 
-> `filters` accept a query string. You can find about the filter syntax on [our dedicated page](/reference/features/filtering.md).
+> `filter` accepts a query string. You can find about the filter syntax on [our dedicated page](/reference/features/filtering_and_faceted_search.md).
```
Comment (bidoubiwa) [inline reference/api/search.md:133]: Same as previous comment, not sure I understand

File: reference/api/search.md [hunk 8]
```diff
@@ -123,16 +122,15 @@ This route should only be used when no API key is required. If an API key is req
...
 | **[offset](/reference/features/search_parameters.md#offset)**                                 | Number of documents to skip                                                                     |     `0`       |
 | **[limit](/reference/features/search_parameters.md#limit)**                                   | Maximum number of documents returned                                                            |     `20`      |
-| **[filters](/reference/features/search_parameters.md#filters)**                               | Filter queries by an attribute value                                                            |    `null`     |
-| **[facetFilters](/reference/features/search_parameters.md#facet-filters)** | Facet names and values to filter on                                                                                |    `null`     |
+| **[filter](/reference/features/search_parameters.md#filter)**                               | Filter queries by an attribute value                                                            |    `null`     |
```
Comment (bidoubiwa) [inline reference/api/search.md:125]: > Filter queries by an attribute value

Not sure this is accurate. maybe `an attribute sorting condition` ?

File: reference/api/filterable_attributes.md (hunk 1 above)
Comment (bidoubiwa) [reply reference/api/filterable_attributes.md:37]: I did not know lazy index also worked on `POST /settings` 🤯

File: reference/features/filtering_and_faceted_search.md [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+Once you have configured `filterableAttributes`, you can start using [the `filter` search parameter](/reference/features/search_parameters#filter). Search parameters are added to a query when a user searches for a document.
+
+`filter` expects a **filter expression** containing one or more **conditions**. A filter expression can be written as a string, as an array, or as a mix of both.
+
+::: warning
+`filter` is a new search parameter introduced in MeiliSearch v0.21. It combines functionality previously present in v0.20's `filters` and `facetFilters`.
+
+Both `filters` and `facetFilters` have been removed from v0.21.
+:::
```
Comment (bidoubiwa) [inline reference/features/filtering_and_faceted_search.md:54]: Since we are not in v1 we do not provide support for the previous MeilISearch versions. To avoid additional maintaining I think it should be removed.
```suggestion
```

File: reference/features/filtering_and_faceted_search.md [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+
+`(genre = horror OR genre = comedy) AND release_date > 795484800`
+
+Failing to add these parentheses will cause the same query to be parsed as:
+
+`genre = horror OR (genre = comedy AND release_date > 795484800)`
+
+Translated into English, the above expression will only return comedies released after March 1995 or horror movies regardless of their `release_date`.
+:::
+
```
Comment (bidoubiwa) [inline reference/features/filtering_and_faceted_search.md:120]: You can add that when using `GET /search` only the expression with strings works.
So not arrays and no mixed

File: reference/features/filtering_and_faceted_search.md [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+#### Creating filter expressions with arrays
+
+Array expressions establish logical connectives by nesting arrays of strings. Array expressions can have a maximum depth of two—arrays with three or more levels of nesting will throw an error.
```
Comment (bidoubiwa) [inline reference/features/filtering_and_faceted_search.md:123]:
```suggestion
Array expressions establish logical connectives by nesting arrays of strings. Array expressions can have a maximum depth of two—arrays, with three or more levels of nesting will throw an error.
```

File: reference/features/filtering_and_faceted_search.md [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+rating >= 3 AND (NOT director = "Tim Burton")
+```
+
+Querying on `Planet of the Apes`, the above example results in the following command:
+
+<CodeSamples id="filtering_guide_4" />
+
+## Faceted search
+
+**Faceted search** is a feature provided out-of-the-box by MeiliSearch. Faceting allows classifying search results into categories or **facets**. Facets are essentially a type of filter.
```
Comment (bidoubiwa) [inline reference/features/filtering_and_faceted_search.md:222]:
```suggestion
**Faceted search** is a sub-filter provided by the filter feature. It was briefly discussed in the previous category. As it is a common feature used in search, the following will explain in depth what it is and how to use it.

Faceting allows classifying search results into categories or **facets**. Facets are essentially a type of filter.

example: In my clothes dataset I want all paints that are either bleu or yellow.
```

File: reference/features/filtering_and_faceted_search.md [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+
+Faceted search provides users with a quick way to narrow down search results by selecting categories relevant to what they are looking for. A faceted navigation system is an **intuitive interface to display and navigate through content**. Facets are used in the UI as filters which users can apply to refine the results in real-time.
+
+This is common in ecommerce sites like Amazon: when users perform a search, they are presented not only with a list of results, but also with a list of facets which you can see on the sidebar in the image below:
+
+![Amazon UI](/faceted-search/facets-amazon.png)
+
+### Filters or facets
+
+Since MeiliSearch v0.21, there is no difference between facets and filters. They function and behave in the exact same way. The question of whether something is a filter or a facet is mostly one pertaining to UX and UI design.
```
Comment (bidoubiwa) [inline reference/features/filtering_and_faceted_search.md:232]: This need to be reworded as facets are a sub-category of filters

**Review by bidoubiwa — CHANGES_REQUESTED** [review]
Mainly I would suggest rewording on the fact that facets are a sub-category of filter to avoid confusing the user. 

For the moment it seems that we are telling them "to use facets you have to use the `filter` query, but facet is different than filter"

File: reference/features/filtering_and_faceted_search.md (hunk 13 above)
Comment (gmourier) [reply reference/features/filtering_and_faceted_search.md:232]: Rather than a sub-category, I think we should say that facetting is a way to filter from a user experience point of view. The term sub category doesn't seem very clear to me for a user.

Also, do we want to mention the version number?

File: learn/what_is_meilisearch/features.md [hunk 14]
```diff
@@ -32,15 +32,15 @@ Read more about synonyms in [this dedicated guide](/reference/features/synonyms.
 
 ## Filters
 
-Meilisearch allows you to define [filters](/reference/features/filtering.md) so you can filter through the results based on criteria.
+Meilisearch allows you to define [filters](/reference/features/filtering_and_faceted_search.md) so you can filter through the results based on criteria.
```
[the remark this answers is not in the document]
Comment (guimachiavelli) [reply learn/what_is_meilisearch/features.md:35]: Yeah, criteria seems a bit vague and just hanging off the end of this sentence. 

I'm not sure I like `sorting condition` because it introduces a concept that is not really expanded upon in the rest of the page. How do you feel about `user-defined criteria`?

```suggestion
MeiliSearch allows you to define [filters](/reference/features/filtering_and_faceted_search.md) so you can filter through the results based on user-defined criteria.
```

File: learn/what_is_meilisearch/features.md (hunk 4 above)
Comment (guimachiavelli) [reply learn/what_is_meilisearch/features.md:43]: I like that idea, though I think the sentence stating the name of the feature (that is, `This feature is called **placeholder search**`) should be together with the main paragraph.

```suggestion
If you make a search without inputting any query words, MeiliSearch will return all the documents in that index sorted by its [custom ranking rules](/reference/features/settings.md#custom-ranking-rule). This feature is called **placeholder search**.

Placeholder searches are particularly effective when used with other features such as [faceting or filtering](/reference/features/filtering_and_faceted_search.md#filters-or-facets), which allow users to narrow their searches and browse by category.
```

File: reference/api/search.md (hunk 8 above)
Comment (guimachiavelli) [reply reference/api/search.md:125]: As I mentioned in another reply, I'm not a fan of `sorting condition` because I'm afraid it only makes sense to a fairly advanced user. As in, a beginner or a user only somewhat familiar with MeiliSearch might look at that and go "WTH is a sorting condition? I want to filter, not sort stuff."

What do you think about:

1. `Filter results using an attribute`

or

2. `Filter results using field attributes and values`

File: reference/features/filtering_and_faceted_search.md [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+
+You can use MeiliSearch's filters to refine search results and create faceted search interfaces.
+
+## Configuring filters
+
+Filters use [document fields](/learn/core_concepts/documents.md#fields) to establish filtering criteria.
+
+To use a document field as a filter, you must first add its attribute to the [`filterableAttributes` index setting](/reference/api/filterable_attributes.md).
+
+**This step is mandatory and cannot be done at search time.** Filters need to be properly processed and prepared by MeiliSearch before they can be used, an operation that takes as much time as indexing all your documents.
```
[the remark this answers is not in the document]
Comment (guimachiavelli) [reply reference/features/filtering_and_faceted_search.md:11]:
```suggestion
**This step is mandatory and cannot be done at search time.** Filters need to be properly processed and prepared by MeiliSearch before they can be used.

Updating `filterableAttributes` requires recreating the entire index. This may take a significant amount of time depending on your dataset size.
```
```

File: reference/features/search_parameters.md (hunk 3 above)
Comment (guimachiavelli) [reply reference/features/search_parameters.md:141]: Oops! Thanks for that!

File: reference/features/filtering_and_faceted_search.md (hunk 11 above)
Comment (guimachiavelli) [reply reference/features/filtering_and_faceted_search.md:118]: This comma would split the subject of the sentence in two (`arrays with three or more levels of nesting` would become a lonely `arrays` followed by `with three or more levels of nesting`), making it a bit confusing.

Just so we're clear: it's hard to see if you're only reviewing the diff, but that dash is not a hyphen, it's an em dash which in this case works like a mix between parentheses and a colon (`:`).

https://deploy-preview-1041--distracted-wiles-3edfc5.netlify.app/reference/features/filtering_and_faceted_search.html#creating-filter-expressions-with-arrays

File: reference/api/filterable_attributes.md (hunk 1 above)
Comment (guimachiavelli) [reply reference/api/filterable_attributes.md:37]: I prefer the shorter version, with minor cosmetic changes:

```suggestion
Update an index's [filterable attributes list](/reference/features/filtering_and_faceted_search.md). This will re-index all documents in the index.
```

File: reference/features/filtering_and_faceted_search.md [hunk 16]
```diff
@@ -0,0 +1,318 @@
+# Filtering and faceted search
+
+You can use MeiliSearch's filters to refine search results and create faceted search interfaces.
```
[the remark this answers is not in the document]
Comment (guimachiavelli) [reply reference/features/filtering_and_faceted_search.md:3]: I really like your second paragraph, but would prefer to keep the first as it is, if that's ok with you. 

My intention would be to:

1. use the first paragraph to give a brief, general and fairly non-technical overview of what users can do with filters;
2. and then use the second paragraph to hammer down the difference between filtering and faceting.

```suggestion
You can use MeiliSearch's filters to refine search results and create faceted search interfaces.

While filters are more general conditions and comparisons (e.g. `age > 18`), faceted search is a type of filter that focuses on categorizing results (e.g. all documents where `color` is `red` or `blue`).
```

File: reference/features/filtering_and_faceted_search.md (hunk 13 above)
Comment (guimachiavelli) [reply reference/features/filtering_and_faceted_search.md:227]: > Rather than a sub-category, I think we should say that facetting is a way to filter from a user experience point of view. The term sub category doesn't seem very clear to me for a user.

I agree with this and it was probably the biggest challenge in the whole article: as far as I understand it, facets are not a different _type_ of filter, they are a different _use_ of filters.

**Comment by guimachiavelli** [comment]
@bidoubiwa Would you be comfortable with saying that a facet is a specific use-case of filters, instead of a subcategory/type?

My problem with saying subcategory is that in the context of software documentation it implies similar but different implementation or usage, when in this case the same filter expression can be used for a facet and for a filter. The only thing that counts when deciding whether something is one or the other is the interface, no?

File: reference/features/filtering_and_faceted_search.md (hunk 12 above)
Comment (guimachiavelli) [reply reference/features/filtering_and_faceted_search.md:217]: Coming from my attempt to clarify that facets are a specific usage of filters, based mostly on how filter conditions/criteria are displayed:

```suggestion
MeiliSearch filters can be used to build **faceted search** interfaces, a specific type of interface that allows a user to refine search results based on broad categories or **facets**. For example, a clothing webshop can use faceted search to allow users to easily explore items in a specific color or belonging to a specific brand.
```

File: reference/features/filtering_and_faceted_search.md (hunk 13 above)
Comment (guimachiavelli) [reply reference/features/filtering_and_faceted_search.md:229]:
```suggestion
### Filters or facets

In MeiliSearch, facets are a specific use-case of filters. The question of whether something is a filter or a facet is mostly one pertaining to UX and UI design.
```

File: reference/features/filtering_and_faceted_search.md [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+        "Horror",
+        "Mystery"
+      ],
+      "overview": "Husband and wife Gabe and Adelaide Wilson take their […]",
+  },
+  …
+]
+```
+
+If you want to filter results based on the `director` and `genre` attributes, you must add them to the [`filterableAttributes` list](/reference/api/filterable_attributes.md):
```
Comment (gmourier) [inline reference/features/filtering_and_faceted_search.md:44]:
```suggestion
If you want to filter results based on the `director` and `genres` attributes, you must add them to the [`filterableAttributes` list](/reference/api/filterable_attributes.md):
```

File: reference/features/filtering_and_faceted_search.md [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+- `NOT` only returns documents that do not satisfy a condition : `NOT genres = horror`
+- `AND` operates by connecting two conditions and only returns documents that satisfy both of them: `genres = horror AND director = 'Jordan Peele'`
+- `OR` connects two conditions and returns results that satisfy at least one of them: `genres = horror OR genres = comedy`
+
+::: tip
+String expressions are read left to right. `NOT` takes precedence over `AND` and `AND` takes precedence over `OR`. You can use parentheses to ensure expressions are correctly parsed.
+
+For instance, if you want your results to only include `comedy` and `horror` movies released after March 1995, the parentheses in the following query are mandatory:
+
+`(genre = horror OR genre = comedy) AND release_date > 795484800`
```
Comment (gmourier) [inline reference/features/filtering_and_faceted_search.md:114]:
```suggestion
`(genres = horror OR genres = comedy) AND release_date > 795484800`
```

File: reference/features/filtering_and_faceted_search.md [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+
+Failing to add these parentheses will cause the same query to be parsed as:
+
+`genre = horror OR (genre = comedy AND release_date > 795484800)`
```
Comment (gmourier) [inline reference/features/filtering_and_faceted_search.md:118]:
```suggestion
`genres = horror OR (genres = comedy AND release_date > 795484800)`
```

File: reference/api/settings.md [hunk 20]
```diff
@@ -67,7 +67,7 @@ List the settings.
     "exactness",
     "desc(release_date)"
   ],
-  "attributesForFaceting": ["genres"],
+  "filterableAttributes": ["genres"],
   "distinctAttribute": null,
   "searchableAttributes": ["title", "description", "genre"],
```
Comment (gmourier) [inline reference/api/settings.md:72]:
```suggestion
  "searchableAttributes": ["title", "description", "genres"],
```

**Review by gmourier — CHANGES_REQUESTED** [review]

**Review by gmourier — APPROVED** [review]
🚀  Thanks @guimachiavelli !
